appserver/web/web-ha/src/main/java/org/glassfish/web/ha/authenticator/HASingleSignOnEntry.java

```diff
@@ -180,7 +183,25 @@ private Principal parse(byte[] pbytes) {
             bais = new ByteArrayInputStream(pbytes);
             bis = new BufferedInputStream(bais);
             ois = ioUtils.createObjectInputStream(bis, true, this.getClass().getClassLoader());
-            return (Principal)ois.readObject();
+
+            if (appClassLoader == null) {
+                return (Principal) ois.readObject();
+            }
+
+            try {
+                return (Principal) ois.readObject();
+            }
+            catch (ClassNotFoundException e) {
+                closeSafely(bis);
+                closeSafely(ois);
+                closeSafely(ois);
```

Comment:
`ois` is double-closed here, shouldn't it be `bais`?

Reply:
That was a typo indeed. I wonder if those inner streams really need to be closed explicitly. Specially closing the byte array stream doesn't do anything really, and the outer one should propagate to the inner ones anyway.

But I wanted, for now, to keep the code as close as possible to the original.